File: product.cpp
#include "product.hpp"
#include <iostream>
#include <iomanip>
#include <ctime>
#include <cmath>
#include <sstream>

Product::Product()
	: m_name("")
	, m_weight(0.0)
	, m_supplier("")
	, m_purchasePrice(0.0)
	, m_expiryDate(0)
	, m_receiptDate(0)
{
}

Product::Product(const std::string& name, double weight, const std::string& supplier,
				 double purchasePrice, std::time_t expiryDate, std::time_t receiptDate)
	: m_name(name)
	, m_weight(weight)
	, m_supplier(supplier)
	, m_purchasePrice(purchasePrice)
	, m_expiryDate(expiryDate)
	, m_receiptDate(receiptDate)
{
}

// Конструктор копирования
Product::Product(const Product& other)
	: m_name(other.m_name)
	, m_weight(other.m_weight)
	, m_supplier(other.m_supplier)
	, m_purchasePrice(other.m_purchasePrice)
	, m_expiryDate(other.m_expiryDate)
	, m_receiptDate(other.m_receiptDate)
{
}

void Product::AddProduct(double quantity)
{
	if (quantity > 0.0)
	{
		m_weight += quantity;
	}
}

void Product::RemoveProduct(double quantity)
{
	if (quantity > 0.0 && quantity <= m_weight)
	{
		m_weight -= quantity;
	}
}

bool Product::CheckExpiryDate() const
{
	std::time_t currentTime = std::time(nullptr);
	return currentTime < m_expiryDate;
}

int Product::GetDaysUntilExpiry() const
{
	std::time_t currentTime = std::time(nullptr);
	double secondsDiff = std::difftime(m_expiryDate, currentTime);
	int daysDiff = static_cast<int>(secondsDiff / (60 * 60 * 24));
	return daysDiff;
}

void Product::PrintInfo() const
{
	std::cout << "=== Информация о продукте ===" << std::endl;
	std::cout << "Название: " << m_name << std::endl;
	std::cout << "Вес: " << std::fixed << std::setprecision(2) << m_weight << " кг" << std::endl;
	std::cout << "Поставщик: " << m_supplier << std::endl;
	std::cout << "Цена закупки: " << m_purchasePrice << " руб." << std::endl;
	
	// Вывод даты поступления (использование std::string вместо массива char)
	std::string receiptDateStr(100, '\0');
	std::strftime(&receiptDateStr[0], receiptDateStr.size(), "%d.%m.%Y", std::localtime(&m_receiptDate));
	receiptDateStr.resize(std::strlen(receiptDateStr.c_str()));  // Обрезаем до реальной длины
	std::cout << "Дата поступления: " << receiptDateStr << std::endl;
	
	// Вывод срока годности (использование std::string вместо массива char)
	std::string expiryDateStr(100, '\0');
	std::strftime(&expiryDateStr[0], expiryDateStr.size(), "%d.%m.%Y", std::localtime(&m_expiryDate));
	expiryDateStr.resize(std::strlen(expiryDateStr.c_str()));  // Обрезаем до реальной длины
	std::cout << "Срок годности: " << expiryDateStr << std::endl;
	
	// Проверка срока годности
	bool isExpired = !CheckExpiryDate();
	std::cout << "Статус срока годности: " << (isExpired ? "Истек" : "Действителен") << std::endl;
	
	if (!isExpired)
	{
		std::cout << "Дней до истечения: " << GetDaysUntilExpiry() << std::endl;
	}
	
	std::cout << "=============================" << std::endl;
}

void Product::SetPurchasePrice(double price)
{
	if (price >= 0.0)
	{
		m_purchasePrice = price;
	}
}

std::string Product::GetName() const
{
	return m_name;
}

double Product::GetWeight() const
{
	return m_weight;
}

std::string Product::GetSupplier() const
{
	return m_supplier;
}

double Product::GetPurchasePrice() const
{
	return m_purchasePrice;
}

std::time_t Product::GetExpiryDate() const
{
	return m_expiryDate;
}

std::time_t Product::GetReceiptDate() const
{
	return m_receiptDate;
}

void Product::SetName(const std::string& name)
{
	m_name = name;
}

void Product::SetWeight(double weight)
{
	if (weight >= 0.0)
	{
		m_weight = weight;
	}
}

void Product::SetSupplier(const std::string& supplier)
{
	m_supplier = supplier;
}

void Product::SetExpiryDate(std::time_t expiryDate)
{
	m_expiryDate = expiryDate;
}

void Product::SetReceiptDate(std::time_t receiptDate)
{
	m_receiptDate = receiptDate;
}

// Реализация перегруженных операторов для Product
Product& Product::operator+=(double quantity)
{
	if (quantity > 0.0)
	{
		this->m_weight += quantity;  // Использование this для доступа к полю
	}
	return *this;  // Возврат ссылки на текущий объект
}

bool Product::operator==(const Product& other) const
{
	return this->m_name == other.m_name;  // Использование this для сравнения
}

// Реализация дружественной функции: сравнение продуктов по цене закупки
bool CompareProductsByPrice(const Product& p1, const Product& p2)
{
	// Дружественная функция имеет доступ к приватным полям
	return p1.m_purchasePrice < p2.m_purchasePrice;
}

File: lab4/Product.java
import java.text.SimpleDateFormat;
import java.util.Date;

public class Product {
    // Приватные поля
    private String name;              // Название продукта
    private double weight;            // Вес
    private String supplier;          // Поставщик
    private double purchasePrice;     // Цена закупки
    private long expiryDate;          // Срок годности (время окончания срока в миллисекундах)
    private long receiptDate;         // Дата поступления (в миллисекундах)

    // Конструктор по умолчанию
    public Product() {
        this.name = "";
        this.weight = 0.0;
        this.supplier = "";
        this.purchasePrice = 0.0;
        this.expiryDate = 0;
        this.receiptDate = 0;
    }

    // Конструктор с параметрами
    // Примечание: expiryDate и receiptDate принимаются в миллисекундах (как в Java Date)
    public Product(String name, double weight, String supplier,
                   double purchasePrice, long expiryDate, long receiptDate) {
        this.name = name;
        this.weight = weight;
        this.supplier = supplier;
        this.purchasePrice = purchasePrice;
        this.expiryDate = expiryDate;
        this.receiptDate = receiptDate;
    }

    // Добавить продукт (увеличить вес)
    public void addProduct(double quantity) {
        if (quantity > 0.0) {
            this.weight += quantity;
        }
    }

    // Удалить продукт (уменьшить вес)
    public void removeProduct(double quantity) {
        if (quantity > 0.0 && quantity <= weight) {
            this.weight -= quantity;
        }
    }

    // Проверка срока годности (true - если срок не истек)
    public boolean checkExpiryDate() {
        long currentTime = System.currentTimeMillis();
        return currentTime < expiryDate;
    }

    // Получить количество дней до истечения срока
    public int getDaysUntilExpiry() {
        long currentTime = System.currentTimeMillis();
        long diffMillis = expiryDate - currentTime;
        int daysDiff = (int)(diffMillis / (1000 * 60 * 60 * 24));
        return daysDiff;
    }

    // Вывести информацию о продукте
    public void printInfo() {
        System.out.println("=== Информация о продукте ===");
        System.out.println("Название: " + name);
        System.out.printf("Вес: %.2f кг%n", weight);
        System.out.println("Поставщик: " + supplier);
        System.out.println("Цена закупки: " + purchasePrice + " руб.");
        
        // Вывод даты поступления
        SimpleDateFormat dateFormat = new SimpleDateFormat("dd.MM.yyyy");
        if (receiptDate > 0) {
            System.out.println("Дата поступления: " + dateFormat.format(new Date(receiptDate)));
        } else {
            System.out.println("Дата поступления: не указана");
        }
        
        // Вывод срока годности
        if (expiryDate > 0) {
            System.out.println("Срок годности: " + dateFormat.format(new Date(expiryDate)));
        } else {
            System.out.println("Срок годности: не указан");
        }
        
        // Проверка срока годности
        boolean isExpired = !checkExpiryDate();
        System.out.println("Статус срока годности: " + (isExpired ? "Истек" : "Действителен"));
        
        if (!isExpired) {
            System.out.println("Дней до истечения: " + getDaysUntilExpiry());
        }
        
        System.out.println("=============================");
    }

    // Установить цену закупки
    public void setPurchasePrice(double price) {
        if (price >= 0.0) {
            this.purchasePrice = price;
        }
    }

    // Геттеры
    public String getName() {
        return name;
    }

    public double getWeight() {
        return weight;
    }

    public String getSupplier() {
        return supplier;
    }

    public double getPurchasePrice() {
        return purchasePrice;
    }

    public long getExpiryDate() {
        return expiryDate;
    }

    public long getReceiptDate() {
        return receiptDate;
    }

    // Сеттеры
    public void setName(String name) {
        this.name = name;
    }

    public void setWeight(double weight) {
        if (weight >= 0.0) {
            this.weight = weight;
        }
    }

    public void setSupplier(String supplier) {
        this.supplier = supplier;
    }

    public void setExpiryDate(long expiryDate) {
        this.expiryDate = expiryDate;
    }

    public void setReceiptDate(long receiptDate) {
        this.receiptDate = receiptDate;
    }
}

